src/utils/splitAtom.ts

```diff
@@ -36,77 +36,101 @@ export function splitAtom<Item, Key>(
   return memoizeAtom(
     () => {
       type ItemAtom = PrimitiveAtom<Item> | Atom<Item>
-      // TODO we should revisit this for a better solution than refAtom
-      const refAtom = atom(
-        () => ({} as { atomList?: ItemAtom[]; keyList?: Key[] })
-      )
-      const read = (get: Getter) => {
-        const ref = get(refAtom)
-        const nextAtomList: Atom<Item>[] = []
-        const nextKeyList: Key[] = []
-        get(arrAtom).forEach((item, index) => {
+      type Mapping = {
+        atomList: ItemAtom[]
+        keyList: Key[]
+      }
+      const mappingCache = new WeakMap<Item[], Mapping>()
+      const getMapping = (arr: Item[], prev?: Item[]) => {
+        let mapping = mappingCache.get(arr)
+        if (mapping) {
+          return mapping
+        }
+        const prevMapping = prev && mappingCache.get(prev)
+        const atomList: Atom<Item>[] = []
+        const keyList: Key[] = []
+        arr.forEach((item, index) => {
           const key = keyExtractor
             ? keyExtractor(item)
             : (index as unknown as Key)
-          nextKeyList[index] = key
-          const cachedAtom = ref.atomList?.[ref.keyList?.indexOf(key) ?? -1]
+          keyList[index] = key
+          const cachedAtom =
+            prevMapping &&
+            prevMapping.atomList[prevMapping.keyList.indexOf(key) ?? -1]
           if (cachedAtom) {
-            nextAtomList[index] = cachedAtom
+            atomList[index] = cachedAtom
             return
           }
+          // TODO we should revisit this for a better solution than refAtom
+          const itemRefAtom = atom(() => ({} as { prev?: Item }))
           const read = (get: Getter) => {
-            const index = ref.keyList?.indexOf(key) ?? -1
-            if (
-              index === -1 &&
-              typeof process === 'object' &&
-              process.env.NODE_ENV !== 'production'
-            ) {
-              console.warn(
-                'splitAtom: array index out of bounds, returning undefined',
-                atom
-              )
+            const itemRef = get(itemRefAtom)
+            const ref = get(refAtom)
+            const arr = get(arrAtom)
+            const mapping = getMapping(arr, ref.prev)
+            const index = mapping.keyList.indexOf(key) ?? -1
+            if (index < 0 || index >= arr.length) {
+              if ('prev' in itemRef) {
+                // returning a stale value to avoid errors for use cases such as react-spring
+                return itemRef.prev
```

Comment:
This PR also handles the useSpring use case, which needs a stale value. Amazing!